README.md
# duperagent [![Build Status](https://travis-ci.org/akoenig/duperagent.svg?branch=master)](https://travis-ci.org/akoenig/duperagent)

Personal convenience wrapper around [superagent](https://github.com/visionmedia/superagent).

## Installation

```
npm install --save duperagent
```

## API

### get(url [, options], callback)

```js
var duperagent = require('duperagent')();

duperagent.get('http://api.host.tld/v1/search.json', function onResponse (err, result) {
    if (err) {
        return console.error(err.message);
    }

    console.log(result);
});
```

### post(url [, options], callback)

```js
duperagent.post('http://api.host.tld/v1/insert.json', {body: {foo: 'bar'}}, function onResponse (err) {
    if (err) {
        return console.error(err.message);
    }

    console.log(result);
})
```

### put(url [, options], callback)

```js
duperagent.put('http://api.host.tld/v1/update.json', {body: {foo: 'zoo'}}, function onResponse (err) {
    if (err) {
        return console.error(err.message);
    }

    console.log(result);
})
```

### del(url [, options], callback)

```js
duperagent.del('http://api.host.tld/v1/delete.json', {query: {id: 1}, function onResponse (err) {
    if (err) {
        return console.error(err.message);
    }

    console.log(result);
})
```

## Author

Copyright 2015, [André König](http://iam.andrekoenig.info) (andre.koenig@posteo.de)


CHANGELOG.md
# 1.0.0 (2015-03-30)

  * 9271c1d Core functionality for `GET`, `POST`, `PUT` and `DELETE`.


specs/index.spec.js
/*
 * duperagent
 *
 * Copyright(c) 2015 André König <andre.koenig@posteo.de>
 * MIT Licensed
 *
 */

/**
 * @author André König <andre.koenig@posteo.de>
 *
 */

'use strict';

var expect = require('expect.js');

var mock = require('./mock');
var duperagent = require('..')();

describe('DuperAgent', function suite () {

    const ENDPOINT = 'http://localhost:9090';

    beforeEach(mock.boot);
    afterEach(mock.shutdown);

    it('should be able to perform a HTTP GET request', function test (done) {

        var params = {};

        function onResponse (err, result) {
            expect(err).to.be(null);

            expect(result.length).not.to.be(0);

            done();
        }

        params.query = {foo: 'bar'};

        duperagent.get(ENDPOINT + '/get', params, onResponse);
    });

    it('should be able to perform a HTTP POST request', function test (done) {
        var params = {};

        function onResponse (err, result) {
            expect(err).to.be(null);

            expect(result.name).to.be('André');

            done();
        }

        params.query = {action: 'insert'};
        params.body = {foo: 'bar'};

        duperagent.post(ENDPOINT + '/post', params, onResponse);
    });

    it('should be able to perform a HTTP PUT request', function test (done) {
        var params = {};

        function onResponse (err, result) {
            expect(err).to.be(null);

            expect(result.name).to.be('König');

            done();
        }

        params.body = {foo: 'bar'};

        duperagent.put(ENDPOINT + '/put', params, onResponse);
    });

    it('should be able to perform a HTTP DELETE request', function test (done) {
        var params = {};

        function onResponse (err) {
            expect(err).to.be(null);

            done();
        }

        duperagent.del(ENDPOINT + '/delete', params, onResponse);
    });

});


gulpfile.js
/*
 * duperagent
 *
 * Copyright(c) 2015 André König <andre.koenig@posteo.de>
 * MIT Licensed
 *
 */

/**
 * @author André König <andre.koenig@posteo.de>
 *
 */

'use strict';

var path = require('path');

var gulp = require('gulp');
var jshint = require('gulp-jshint');
var jscs = require('gulp-jscs');
var mocha = require('gulp-mocha');
var sequence = require('run-sequence');

var paths = {};

paths.sources = [
    path.join(__dirname, '*.js'),
    path.join(__dirname, 'lib', '*.js'),
    path.join(__dirname, 'specs', '**', '*.spec.js')
];

paths.specs = [path.join(__dirname, 'specs', '**', '*.spec.js')];

gulp.task('specs', function specs () {
    return gulp.src(paths.specs, {read: false})
        .pipe(mocha({reporter: 'nyan'}));
});

gulp.task('lint', function lint () {
    return gulp.src(paths.sources)
        .pipe(jshint())
        .pipe(jshint.reporter('jshint-stylish'))
        .pipe(jshint.reporter('fail'));
});

gulp.task('checkstyle', function checkstyle () {
    return gulp.src(paths.sources)
        .pipe(jscs());
});

gulp.task('default', function defaultTask (callback) {
    return sequence('lint', 'checkstyle', 'specs', callback);
});


lib/index.js
/*
 * duperagent
 *
 * Copyright(c) 2015 André König <andre.koenig@posteo.de>
 * MIT Licensed
 *
 */

/**
 * @author André König <andre.koenig@posteo.de>
 *
 */

'use strict';

var debug = require('debug')('duperagent');
var mandatory = require('mandatory');
var superagent = require('superagent');
var VError = require('verror');

module.exports = function instantiate () {
    var agent = new DuperAgent();

    return {
        get: agent.fireGET.bind(agent),
        post: agent.firePOST.bind(agent),
        put: agent.firePUT.bind(agent),
        del: agent.fireDELETE.bind(agent)
    };
};

function DuperAgent () {}

DuperAgent.prototype.$request = function $request (verb, uri, params, callback) {

    function onResponse (err, res) {
        if (err) {
            debug('Request failed: %s', err.toString());
            return callback(new VError(err, 'failed to communicate with the respective server'));
        }

        if (res.status < 200 || res.status >= 400) {
            debug('Server sent an error response: %d', res.status);

            return callback(new VError('Server sent an error response: %s', res.status));
        }

        if (res.text) {
            try {
                return callback(null, JSON.parse(res.text));
            } catch (err) {
                debug('Response was not parsable: %s', err.toString());
                return callback(new VError(err, 'Unable to parse response body'));
            }
        }

        callback(null);
    }

    if (typeof params === 'function') {
        callback = params;
        params = {};
    }

    params.query = params.query || {};

    superagent[verb.toLowerCase()](uri)
        .unset('User-Agent')
        .query(params.query)
        .send(params.body)
        .end(onResponse);
};

DuperAgent.prototype.$validate = function $validate (uri, params, callback) {
    if (typeof params === 'function') {
        callback = params;
        params = {};
    }

    mandatory(uri).is('string', 'Please provide a proper uri');
    mandatory(params).is('object', 'Please provide a proper parameter object');
    mandatory(callback).is('function', 'Please provide a proper callback function');
};

DuperAgent.prototype.fireGET = function fireGET (uri, params, callback) {

    function onResponse (err, body) {
        if (err) {
            return callback(new VError(err, 'failed to perform HTTP GET request'));
        }

        callback(null, body);
    }

    this.$validate(uri, params, callback);

    this.$request('GET', uri, params, onResponse);
};

DuperAgent.prototype.firePOST = function firePOST (uri, params, callback) {
    function onResponse (err, body) {
        if (err) {
            return callback(new VError(err, 'failed to perform HTTP POST request'));
        }

        callback(null, body);
    }

    this.$validate(uri, params, callback);

    this.$request('POST', uri, params, onResponse);
};

DuperAgent.prototype.firePUT = function firePUT (uri, params, callback) {
    function onResponse (err, body) {
        if (err) {
            return callback(new VError(err, 'failed to perform HTTP PUT request'));
        }

        callback(null, body);
    }

    this.$validate(uri, params, callback);

    this.$request('PUT', uri, params, onResponse);
};

DuperAgent.prototype.fireDELETE = function fireDELETE (uri, params, callback) {
    function onResponse (err, body) {
        if (err) {
            return callback(new VError(err, 'failed to perform HTTP DELETE request'));
        }

        callback(null, body);
    }

    this.$validate(uri, params, callback);

    this.$request('DEL', uri, params, onResponse);
};


specs/mock.js
/*
 * duperagent
 *
 * Copyright(c) 2015 André König <andre.koenig@posteo.de>
 * MIT Licensed
 *
 */

/**
 * @author André König <andre.koenig@posteo.de>
 *
 */

'use strict';

var shmock = require('shmock');

const PORT = 9090;

var mock;

module.exports.boot = function boot () {
    var resource;

    mock = shmock(PORT);

    mock
        .get('/get')
        .query({foo: 'bar'})
        .reply(200, JSON.stringify(['foo', 'bar']));

    mock
        .post('/post')
        .send({foo: 'bar'})
        .query({action: 'insert'})
        .reply(201, JSON.stringify({name: 'André'}));

    mock
        .put('/put')
        .send({foo: 'bar'})
        .reply(200, JSON.stringify({name: 'König'}));

    mock
        .delete('/delete')
        .reply(200, JSON.stringify(''));
};

module.exports.shutdown = function shutdown (callback) {
    mock.close(callback);
};
